### PR TITLE
Improve Go harness prompt state

### PR DESCRIPTION
**Pull request by jhtschultz**
## Summary

- Keep the raw JSON state in the Go prompt while adding a readable ASCII board for the same position.
- Source full-game move history from the Go proxy instead of the per-agent core harness closure history.
- Make GTP coordinate guidance board-size-aware and clarify that final JSON uses lowercase coordinate-only moves.
- Add simple-ko wording and keep strict JSON-only parsing through the shared `core_harness.parse_json_action` helper.
- Rename new proxy state from `previous_move_a1` to `previous_move`, while visualizers still accept old replay data.

## Checklist

- [x] Keep raw JSON state in prompt as the source of metadata, including `komi`.
- [x] Add ASCII board in addition to JSON state.
- [x] Add full-game move history from `GoState.history()`.
- [x] Use full-game history in prompt instead of per-agent `core_harness` history.
- [x] Make coordinate guidance board-size-aware.
- [x] Keep parser strict: JSON `move` is coordinate-only or `PASS`; reject `"B e5"`.
- [x] Replace `previous_move_a1` with `previous_move` for new proxy state.
- [x] Keep visualizers compatible with older `previous_move_a1` replay data.
- [x] Reconcile uppercase board labels with lowercase final JSON wording.
- [x] Add simple-ko wording.

## Full Example Prompt

```text
Let's play Go.

Rules: Tromp-Taylor scoring (area scoring — count stones on the board plus
empty territory enclosed by a single color; all stones are treated as alive).
Komi is given in the game state JSON below. Two differences from standard
Tromp-Taylor: (1) suicide is illegal — you may not place a stone that would
be immediately captured unless it captures enemy stones first, and
(2) positional superko violations end the game as a draw rather than simply
making the move illegal. Simple ko is also illegal: after a single-stone
capture, the opponent may not immediately recapture on that same point.
The game ends when both players pass consecutively.

The current game state JSON is:
{"board_size": 13, "komi": 7.5, "current_player_to_move": "B", "move_number": 1, "previous_move": null, "move_history": [], "ascii_board": "13 +++++++++++++\n12 +++++++++++++\n11 +++++++++++++\n10 +++++++++++++\n 9 +++++++++++++\n 8 +++++++++++++\n 7 +++++++++++++\n 6 +++++++++++++\n 5 +++++++++++++\n 4 +++++++++++++\n 3 +++++++++++++\n 2 +++++++++++++\n 1 +++++++++++++\n   ABCDEFGHJKLMN", "board_grid": [[{"A13": "."}, {"B13": "."}, {"C13": "."}, {"D13": "."}, {"E13": "."}, {"F13": "."}, {"G13": "."}, {"H13": "."}, {"J13": "."}, {"K13": "."}, {"L13": "."}, {"M13": "."}, {"N13": "."}], [{"A12": "."}, {"B12": "."}, {"C12": "."}, {"D12": "."}, {"E12": "."}, {"F12": "."}, {"G12": "."}, {"H12": "."}, {"J12": "."}, {"K12": "."}, {"L12": "."}, {"M12": "."}, {"N12": "."}], [{"A11": "."}, {"B11": "."}, {"C11": "."}, {"D11": "."}, {"E11": "."}, {"F11": "."}, {"G11": "."}, {"H11": "."}, {"J11": "."}, {"K11": "."}, {"L11": "."}, {"M11": "."}, {"N11": "."}], [{"A10": "."}, {"B10": "."}, {"C10": "."}, {"D10": "."}, {"E10": "."}, {"F10": "."}, {"G10": "."}, {"H10": "."}, {"J10": "."}, {"K10": "."}, {"L10": "."}, {"M10": "."}, {"N10": "."}], [{"A9": "."}, {"B9": "."}, {"C9": "."}, {"D9": "."}, {"E9": "."}, {"F9": "."}, {"G9": "."}, {"H9": "."}, {"J9": "."}, {"K9": "."}, {"L9": "."}, {"M9": "."}, {"N9": "."}], [{"A8": "."}, {"B8": "."}, {"C8": "."}, {"D8": "."}, {"E8": "."}, {"F8": "."}, {"G8": "."}, {"H8": "."}, {"J8": "."}, {"K8": "."}, {"L8": "."}, {"M8": "."}, {"N8": "."}], [{"A7": "."}, {"B7": "."}, {"C7": "."}, {"D7": "."}, {"E7": "."}, {"F7": "."}, {"G7": "."}, {"H7": "."}, {"J7": "."}, {"K7": "."}, {"L7": "."}, {"M7": "."}, {"N7": "."}], [{"A6": "."}, {"B6": "."}, {"C6": "."}, {"D6": "."}, {"E6": "."}, {"F6": "."}, {"G6": "."}, {"H6": "."}, {"J6": "."}, {"K6": "."}, {"L6": "."}, {"M6": "."}, {"N6": "."}], [{"A5": "."}, {"B5": "."}, {"C5": "."}, {"D5": "."}, {"E5": "."}, {"F5": "."}, {"G5": "."}, {"H5": "."}, {"J5": "."}, {"K5": "."}, {"L5": "."}, {"M5": "."}, {"N5": "."}], [{"A4": "."}, {"B4": "."}, {"C4": "."}, {"D4": "."}, {"E4": "."}, {"F4": "."}, {"G4": "."}, {"H4": "."}, {"J4": "."}, {"K4": "."}, {"L4": "."}, {"M4": "."}, {"N4": "."}], [{"A3": "."}, {"B3": "."}, {"C3": "."}, {"D3": "."}, {"E3": "."}, {"F3": "."}, {"G3": "."}, {"H3": "."}, {"J3": "."}, {"K3": "."}, {"L3": "."}, {"M3": "."}, {"N3": "."}], [{"A2": "."}, {"B2": "."}, {"C2": "."}, {"D2": "."}, {"E2": "."}, {"F2": "."}, {"G2": "."}, {"H2": "."}, {"J2": "."}, {"K2": "."}, {"L2": "."}, {"M2": "."}, {"N2": "."}], [{"A1": "."}, {"B1": "."}, {"C1": "."}, {"D1": "."}, {"E1": "."}, {"F1": "."}, {"G1": "."}, {"H1": "."}, {"J1": "."}, {"K1": "."}, {"L1": "."}, {"M1": "."}, {"N1": "."}]]}

ASCII board for the same position (X=Black, O=White, +=empty; board labels may be uppercase):
13 +++++++++++++
12 +++++++++++++
11 +++++++++++++
10 +++++++++++++
 9 +++++++++++++
 8 +++++++++++++
 7 +++++++++++++
 6 +++++++++++++
 5 +++++++++++++
 4 +++++++++++++
 3 +++++++++++++
 2 +++++++++++++
 1 +++++++++++++
   ABCDEFGHJKLMN

The full game move history is:
None

You are playing as player Black (B).
It is now your turn. Play your strongest move.
The move MUST be legal.
Your response should include the reasoning that led you to your move, and
conclude with your final move as a JSON formatted as follows:

```json
{
  "move": "<move>" }
```

Where move is the coordinate only (e.g. "a1", "b2", "e5") or "PASS" if you wish to pass.
Coordinates use GTP notation for this 13x13 board: columns are a-h,j-n (the letter "i" is skipped), and rows are 1-13.
The board display may use uppercase labels, but your final JSON move must use the lowercase coordinate only.
Failure to output your final answer in the specified format will result in a loss.
Begin!
```

## Tests

```sh
.venv/bin/python -m pytest tests/envs/open_spiel_env/games/go/harness_test.py -q
.venv/bin/python -m py_compile kaggle_environments/envs/open_spiel_env/games/go/harness.py kaggle_environments/envs/open_spiel_env/games/go/go_proxy.py tests/envs/open_spiel_env/games/go/harness_test.py
git diff --check
```

### CODE REVIEW

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py [hunk 1]
```diff
@@ -83,6 +94,72 @@
...
+    if isinstance(history, list):
+        return " ".join(str(move) for move in history) or "None"
+    return str(history)
+
+
+def _ascii_board_from_state(state: Mapping[str, Any]) -> str:
+    board = state.get("ascii_board")
+    if isinstance(board, str) and board.strip():
+        return board
+    return "Not available in this observation."
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/go/harness.py:160]: Critical: `_ascii_board_from_state` reads `state['ascii_board']`, but the Go proxy's `state_dict()` (go_proxy.py:147-157) does not produce an `ascii_board` field — only `board_size`, `komi`, `current_player_to_move`, `move_number`, `previous_move_a1`, `board_grid`, and optionally `scoring`. So the prompt will always render the literal text `Not available in this observation.` under the heading 'ASCII board for the same position…'. Either (a) add `ascii_board` to GoState.state_dict() (e.g. by formatting the wrapped pyspiel state's `__str__()`), or (b) drop the ASCII-board section from the prompt template. Right now the prompt is advertising context that never exists.

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py [hunk 2]
```diff
@@ -83,6 +94,72 @@
...
+        return len(board_grid)
+    return None
+
+
+def _format_full_move_history(state: Mapping[str, Any]) -> str:
+    history = state.get("move_history")
+    if not history:
+        return "None"
+    if isinstance(history, list):
+        return " ".join(str(move) for move in history) or "None"
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/go/harness.py:152]: Critical regression: `_format_full_move_history` reads `state['move_history']`, but the Go proxy never sets that key (go_proxy.py only exposes `previous_move_a1`, not a full history). Combined with `del move_history` at line ~219 — which discards the engine-supplied `move_history` argument that this harness previously used — the move-history portion of the prompt will always be `None`. The model used to see the entire game so far via the parameter; now it sees nothing. Either populate `move_history` in the proxy's state_dict, or fall back to the `move_history` parameter when the state field is missing instead of `del`-ing it.

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py [hunk 3]
```diff
@@ -22,16 +27,23 @@
...
 empty territory enclosed by a single color; all stones are treated as alive).
-Komi is given in the game state below. Two differences from standard
+Komi is given in the game state JSON below. Two differences from standard
 Tromp-Taylor: (1) suicide is illegal — you may not place a stone that would
 be immediately captured unless it captures enemy stones first, and
 (2) positional superko violations end the game as a draw rather than simply
-making the move illegal. The game ends when both players pass consecutively.
+making the move illegal. Simple ko is also illegal: after a single-stone
+capture, the opponent may not immediately recapture on that same point.
+The game ends when both players pass consecutively.
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/go/harness.py:36]: Minor consistency issue in the prompt: the new 'Simple ko is also illegal: …may not immediately recapture…' sentence reads as if simple-ko moves are merely rejected, but the preceding sentence already says that positional-superko violations (which subsume simple ko in OpenSpiel) end the game as a draw rather than being rejected. A model trying to reconcile these may conclude that simple-ko recapture is safe to attempt because it 'just' makes the move illegal. Consider unifying the wording so the consequence (game ends as a draw) is clearly stated for both simple ko and superko.

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py [hunk 4]
```diff
@@ -83,6 +94,72 @@
...
+        return "the columns shown in the board state"
+    columns = _GTP_COLUMNS[:board_size]
+    if board_size <= 8:
+        return f"a-{columns[-1]}"
+    if board_size == 9:
+        return "a-h,j"
+    return f"a-h,j-{columns[-1]}"
+
+
+def _coordinate_guidance(board_size: int | None) -> str:
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/go/harness.py:121]: Minor: `_GTP_COLUMNS` is 25 characters (a–z minus i). For any board_size > 25 (not currently supported by OpenSpiel Go, which caps at 19, so not exploitable today), `_GTP_COLUMNS[:board_size]` silently truncates and `columns[-1]` returns the wrong column. Worth either clamping/asserting or noting that 19 is the upper bound. Low priority since the engine constrains this in practice.

**Review by kaggle-agent — COMMENTED** [review]
**review-harness**: The prompt now promises an ASCII board and full move history sourced from the proxy's state JSON, but the Go proxy (go_proxy.py state_dict) does not emit either an `ascii_board` or `move_history` key. As a result every Go prompt will render the placeholder 'Not available in this observation.' for the board and 'None' for the move history. Combined with `del move_history` (which throws away the engine-provided history the harness used to use), this is a strict regression: models lose the move-history context they previously had, and the promised ASCII board never appears.

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py (hunk 2 above)
Comment (lipovetz) [reply kaggle_environments/envs/open_spiel_env/games/go/harness.py:152]: Looks like a false alarm that didn't take into account the delta in go_proxy.py

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py (hunk 1 above)
Comment (lipovetz) [reply kaggle_environments/envs/open_spiel_env/games/go/harness.py:160]: Ditto

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py (hunk 4 above)
Comment (lipovetz) [reply kaggle_environments/envs/open_spiel_env/games/go/harness.py:121]: Feel free to ignore

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py (hunk 3 above)
Comment (lipovetz) [reply kaggle_environments/envs/open_spiel_env/games/go/harness.py:36]: Minor but fair point.

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py [hunk 5]
```diff
@@ -45,9 +57,8 @@
 ```
 
 Where move is the coordinate only (e.g. "a1", "b2", "e5") or "PASS" if you wish to pass.
-Coordinates use GTP notation: columns are lowercase letters a-h, j (the letter
-"i" is skipped to avoid confusion with "l"), rows are numbers starting from 1.
-For example on a 9x9 board, columns are a-h,j and rows are 1-9.
+{coordinate_guidance}
+The board display may use uppercase labels, but your final JSON move must use the lowercase coordinate only.
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/go/harness.py]: Why not standardize on either upper or lower?

**Review by lipovetz — APPROVED** [review]
A couple minor things to consider, otherwise LGTM

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py (hunk 2 above)
Comment (jhtschultz) [reply kaggle_environments/envs/open_spiel_env/games/go/harness.py:152]: Confirmed

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py (hunk 1 above)
Comment (jhtschultz) [reply kaggle_environments/envs/open_spiel_env/games/go/harness.py:160]: Confirmed

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py (hunk 3 above)
Comment (jhtschultz) [reply kaggle_environments/envs/open_spiel_env/games/go/harness.py:36]: Updated with more explicit description.

File: kaggle_environments/envs/open_spiel_env/games/go/harness.py (hunk 5 above)
Comment (jhtschultz) [reply kaggle_environments/envs/open_spiel_env/games/go/harness.py]: Agreed, standardized on upper, consistent with GTP.
